### PR TITLE
#407 Rewrite layout/Header to use TypeScript and hooks

### PR DESCRIPTION
**Pull request by guillaumeLamanda**
#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] [CODEOWNERS](./CODEOWNERS) updated (for new stuff)
- [ ] Regression tests added for bug fixes

### Description
In the scope of #407, I rewrite the missing header components

- `<ContentHeader>`
- `<HeaderActionMenu>`
- `<HeaderLabel>`

I have made the changes coherent with https://github.com/spotify/backstage/pull/408

I have also moved the `tsc` build options in the `build` script of `package.json` in `tsconfig.json` for a nicer integration with vscode. I verified the `build` phase passed.

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Thanks, looks good! Some minor comments to fix before merge

File: packages/core/src/layout/HeaderLabel/HeaderLabel.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+type HeaderLabelProps = {
+  label: string;
+  value?: HeaderLabelContentProps['value'];
+  url?: string;
+};
+
+const HeaderLabel: FC<HeaderLabelProps> = ({
+  label,
+  value = '<Unknown>',
```
[the remark this answers is not in the document]
Comment (guillaumeLamanda) [reply packages/core/src/layout/HeaderLabel/HeaderLabel.tsx]: I updated my PR to follow the previous behavior, and I added a test to verify this behavior. 
https://github.com/spotify/backstage/pull/425/commits/7774f3939e3795ccecebb387a36fffa45b2d9c46#diff-350782cc3b15913265ed1de6ce4eb6faR33

File: packages/core/src/layout/ContentHeader/ContentHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+}) => (
+  <Typography variant="h4" className={className} data-testid="header-title">
+    {title}
+  </Typography>
+);
+
+type ContentHeaderProps = {
+  titleComponent?: ComponentType;
+  description?: string;
+} & Pick<DefaultTitleProps, 'title'>;
```
Comment (Rugvip) [inline packages/core/src/layout/ContentHeader/ContentHeader.tsx]: Think I'd prefer the ´title: DefatultTitleProps['title']´ approach, afaik that displays better in error messages and editor popups. Not a blocker though.

**Review by Rugvip — DISMISSED** [review]
Thanks! Looks good, trying to keep up in #407 :grin:

**Comment by stefanalund** [comment]
@guillaumeLamanda great job so far. Can you please rebase and then we will merge?

File: packages/core/src/layout/ContentHeader/ContentHeader.tsx (hunk 2 above)
Comment (guillaumeLamanda) [reply packages/core/src/layout/ContentHeader/ContentHeader.tsx]: I updated this type to use the other approach. 
But doing this, you lose the composition of the types of typescript, which could be redundant if you have more props to pick. Even here, `title?` is repeated 2 times, which is more prone to errors (like forgetting the `?` comma)

**Comment by guillaumeLamanda** [comment]
@stefanalund this is now rebased. 👍

**Review by stefanalund — APPROVED** [review]

**Comment by stefanalund** [comment]
@guillaumeLamanda thanks a lot! 🙏

**Pull request merged**